### DISCUSSION TITLE
Using jq in Mirth to query JSON

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#5542] [General]
Has anyone ever tried to use [jq](https://stedolan.github.io/jq/) (JSON query/proessor) in Mirth?

There seem to be Java wrappers [java-jq](https://github.com/arakelian/java-jq) (embeds native jq libs) or [
jackson-jq](https://github.com/eiiches/jackson-jq) (pure Java jq implementation)

I find it quite handy when you need to define complex message filters: jq could do it in one line of code (like XPath), what makes it interesting to define a filter outside of Mirth (e.g. using a config file or GUI). You could also use jq to extract data from a JSON, e.g. to convert it into a payload for a REST call.

**Comment by jonbartels** [comment]
`jq` is great on the command line. Do you know if the libraries you listed have a lot of transitive dependencies and if those dependencies conflict with anything already on the Mirth classpath?

**Reply by tom08zehn** [reply]
Unfortunately not. I'm not a Java expert, not at all... For the jackson-jq it shows:

```xml
<dependency>
	<groupId>net.thisptr</groupId>
	<artifactId>jackson-jq</artifactId>
	<version>1.0.0-preview.20220705</version>
</dependency>
```

And the [Java API usage](https://github.com/eiiches/jackson-jq/blob/develop/1.x/jackson-jq/src/test/java/examples/Usage.java) shows the following imports:

```java
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.FileSystems;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.IntNode;
import com.fasterxml.jackson.databind.node.TextNode;

import net.thisptr.jackson.jq.BuiltinFunctionLoader;
import net.thisptr.jackson.jq.Expression;
import net.thisptr.jackson.jq.Function;
import net.thisptr.jackson.jq.JsonQuery;
import net.thisptr.jackson.jq.PathOutput;
import net.thisptr.jackson.jq.Scope;
import net.thisptr.jackson.jq.Version;
import net.thisptr.jackson.jq.Versions;
import net.thisptr.jackson.jq.exception.JsonQueryException;
import net.thisptr.jackson.jq.internal.misc.Strings;
import net.thisptr.jackson.jq.module.ModuleLoader;
import net.thisptr.jackson.jq.module.loaders.BuiltinModuleLoader;
import net.thisptr.jackson.jq.module.loaders.ChainedModuleLoader;
import net.thisptr.jackson.jq.module.loaders.FileSystemModuleLoader;
import net.thisptr.jackson.jq.path.Path;
```

Does it help?

**Reply by jonbartels** [reply]
Close. `mvn dependency:tree -Dscope=compile` shows the transitive dependencies, ignoring test dependencies:

```
[INFO] net.thisptr:jackson-jq:bundle:1.0.0-SNAPSHOT
[INFO] +- org.jruby.joni:joni:jar:2.1.43:compile
[INFO] |  \- org.jruby.jcodings:jcodings:jar:1.0.57:compile
[INFO] +- com.fasterxml.jackson.core:jackson-databind:jar:2.13.3:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.13.3:compile
[INFO] \- com.google.auto.service:auto-service-annotations:jar:1.0.1:compile
```

Then looking in MC we can see that MC includes Jackson 2.11. There might be a conflict with Jackson but maybe not, it is worth trying. 

When you grab the jackson-jq JAR wou'll also need the JARs for JRuby too. Try it with only that and see if it works with the Jackson 2.11 already available in MC.

**Comment by tom08zehn** [comment]
@jonbartels Thanks for the details but I changed my mind.

On the one hand, as I'm not a Java expert, it would probably take me hours/days to figure out if this will work... and maybe the answer is that it doesn't work the way I need it.

On the other hand I figured out that Mirth has full XPath support built-in what makes it even more flexible than just JSON query.

For those who are interested, see the source code to get an example: https://github.com/nextgenhealthcare/connect/blob/master/server/src/com/mirth/connect/plugins/datatypes/hl7v2/HL7v2AutoResponder.java#L120

My idea was to select values from fields, partly using a more complex query logic like `if /path/to/a == 'foo' and /path/to/b == 'bar' then return concat(/path/to/c, '-', replace(/path/to/d, 'eggs', 'spam'))` ... something like this, and XPath can perfectly do that.
